CI/CD with Github Actions

N-Storm opened the pull request:
This PR introduces automated builds using GitHub Actions (GA) runners, with support for uploading resulting binaries to the "Releases" section, along with release notes generation. The latter happens when a tag is created, but this can be adjusted to match a specific template or be triggered manually.

Supported configurations:
- **Windows x64**: CUDA 11.0 and higher  
- **Linux x86_64**: CUDA 8.0 and higher  

Building 32-bit binaries and supporting older CUDA versions within GA runners doesn’t seem reasonable to me. The time required to implement and test these would likely outweigh any demand. If possible at all—since, for example, Windows GA runners are only available in two flavors (MSVC 2019 & 2022). The currently supported versions should cover any hardware still useful for factoring.

Currently, I’ve included one target version per **MAJOR.MINOR**, selecting the latest **.PATCHLEVEL** available. In other words, for every **X.Y.z**, I pick one **X.Y** with the highest available **.z**. However, this can be easily configured in the [`.github/workflows/build.yml`](https://github.com/N-Storm/mfaktc/blob/511f1868805aa30b2aa8ae2f785693e766a77cdc/.github/workflows/build.yml). I’ve left comments there with links to available versions.  

It’s also possible to add a release flag so that generic builds run for every version listed, but only those marked with this flag get uploaded to the releases section. This is useful if you prefer not to upload too many releases but still want continuous builds to gather feedback on the build process as further changes are made. Builds that aren’t uploaded can still be downloaded by repository members with write permissions under **Actions → [Build] → Artifacts**, where they remain available for up to 90 days (configurable in the repository settings on the GitHub web interface).

I had to [slightly modify `Makefile.win`](https://github.com/primesearch/mfaktc/commit/a2a45c50d793074dec8594c5651ce410e0dd7dd5#diff-fd95cf91a5934e10c646e04552d16af97ddc71457ff76089a2090626d1d5e683L1) to fix path and quoting issues for CUDA. Instead of using an intermediate `CUDA_DIR`, I switched to using `CUDA_PATH` directly. This should be safe for manual builds as well, but I haven’t tested it. If this change breaks manual builds or is otherwise unacceptable, it can be moved to a build helper script—modifying a local copy of `Makefile.win` during the build while keeping the repository version unchanged.

### Example releases built using this workflow:
- [0.23.2pre4](https://github.com/N-Storm/mfaktc/releases/tag/0.23.2pre4)  
- Recently added **0.24.0** test branch builds as well, if `Makefile.win` changes from this PR are merged into it:  
  [Build log](https://github.com/N-Storm/mfaktc/actions/runs/13499381031/job/37715598754) + [Release link](https://github.com/N-Storm/mfaktc/releases/tag/0.24.0pre1)  

_Note: The 0.24.0 test branch release was built before a fix for the table was committed, so columns with CUDA versions are missing and Compute Capability is misaligned. This has already been fixed in this PR, but I haven’t rebuilt 0.24.0 branch on my fork._  

Feel free to PM me on the forums if you have any questions.

N-Storm commented:
> I'm also thinking we might want this PR to be the patch version bump to 0.23.2, if you wouldn't mind changing at params.h:89 to 0.23.2, then the release can be tagged and be the first automated one (which will also include the log corruption fix)

Thanks! I'll adjust it right now and update this PR soon.

brubsby commented:
Also, fwiw, the changes to Makefile.win had no adverse effect on my personal build environment, so we're presumably good on that front as well.

N-Storm commented:
Hmm... how do you think, maybe I should add my patch which replaces deprecated cudaThreadSetCacheConfig() and cudaThreadSynchronize() calls to the supported cudaDeviceSetCacheConfig() / cudaDeviceSynchronize() respectively to include it in the release?
I've posted about this on same forum thread recently. This are the commits for it: https://github.com/primesearch/mfaktc/commit/e737557b20680cac8628c6550fbee63203359ebb, https://github.com/primesearch/mfaktc/commit/7b536c60f269a57be78aa12653965a6964fef8aa (2nd are just a fix for a small error I've did, which only affects builds with DEBUG_GPU_MATH build option enabled)
It looks like 0.24.0 incorporates similar changes, nearly the same. I've tested those running a Windows for a few days without any issues now. Didn't noticed any difference with my compute 8.9 device, neither it helped to fix issues on RTX 5090. But doesn't hurts at least and might be of use with some other GPUs as well.

N-Storm commented:
Added patch to replace deprecated API calls as well. Let me know if want it out of this PR / v0.23.2, I'll exclude it and force-push the branch without it.

KarlLudwig3485 commented:
Seems to work fine on my setups.
I hope that any builds for CUDA <=11.x you make can run on Windows 7, as it seems a few people still use it on their systems. Otherwise I would be more than happy to continue providing builds for legacy/novelty systems (which I will continue to do for Windows XP and similar).

And I'm glad to report that your changes still work with CUDA 5.0 :)

KarlLudwig3485 commented:
My apologies, I didn't see that the Windows builds only went as far back as CUDA 11.0. And, given that Visual Studio 2022 still supports targeting Windows 7 by default (as far as I'm aware), it should work just fine.

As for CUDA 5.0, I was referring to the changes regarding `cudaThreadSetCacheConfig()` and `cudaThreadSynchronize()` being replaced with `cudaDeviceSetCacheConfig()` / `cudaDeviceSynchronize()`. Interestingly the former were already deprecated as early as CUDA 5.0 (I haven't checked any earlier versions yet).

N-Storm commented:
@KarlLudwig3485 got it. Thanks for the feedback! Didn't thought at first that your comment was about that last addition, which I've included so recently. And it's so unrelated to the PR so I wasn't even thinking about it while reading your comment. :)
Yes, those call was declared deprecated pretty long time ago. But as NVidia explains on the deprecation notice: "Note that this function is deprecated because its name does not reflect its behavior. Its functionality is similar to the non-deprecated function [cudaDeviceSynchronize()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g10e20b05a95f638a4071a655503df25d), which should be used instead.". So I guess those functions could be just a renamed or aliased. With the exception of cudaStreamSynchronize(), which work with the CUDA streams.

tdulcet commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,248 @@
+      - name: Prepare list of release files & release notes
+        id: makeinfo
+        run: |
+          echo "Binary releases (automated builds) below." > RELEASE_NOTES.txt
+          echo "Compute Capability (CC) in the table means minimum and maximum versions supported." >> RELEASE_NOTES.txt
+          echo "CC versions are listed without the dot (i.e. 90 means 9.0 compute capability)." >> RELEASE_NOTES.txt
+          echo >> RELEASE_NOTES.txt
+          echo "File | CUDA version | Compute Capability | Build OS | Compiler version | NVCC version" >> RELEASE_NOTES.txt
+          echo "--- | --- | --- | --- | --- | ---" >> RELEASE_NOTES.txt
+          cat mfaktc-${{ github.ref_name }}-*-cuda*/mfaktc-${{ github.ref_name }}-*-cuda*.txt | sort -Vr >> RELEASE_NOTES.txt
+          echo 'RELEASE_FILES<<EOF' > $GITHUB_OUTPUT
+          ls -1 mfaktc-${{ github.ref_name }}-*-cuda*/mfaktc-${{ github.ref_name }}-*-cuda*.zip | sort -Vr >> $GITHUB_OUTPUT
+          echo 'EOF' >> $GITHUB_OUTPUT
```
```suggestion
          {
            echo "Binary releases (automated builds) below."
            echo "Compute Capability (CC) in the table means minimum and maximum versions supported."
            echo "CC versions are listed without the dot (i.e. 90 means 9.0 compute capability)."
            echo
            echo "File | CUDA version | Compute Capability | Build OS | Compiler version | NVCC version"
            echo "--- | --- | --- | --- | --- | ---"
            sort -Vr mfaktc-${{ github.ref_name }}-*-cuda*/mfaktc-${{ github.ref_name }}-*-cuda*.txt
          } > RELEASE_NOTES.txt
          {
            echo 'RELEASE_FILES<<EOF'
            printf '%s\n' mfaktc-${{ github.ref_name }}-*-cuda*/mfaktc-${{ github.ref_name }}-*-cuda*.zip | sort -Vr
            echo 'EOF'
          } > $GITHUB_OUTPUT
```
Also, removed the useless `cat` and the useless `ls`.

tdulcet commented on .github/workflows/scripts/build_helper.sh:
```diff
@@ -0,0 +1,102 @@
+  exit 2
+fi
+
+eval $(echo -e "CUDA_VER_MAJOR=${CUDA_VERSION[0]}\nCUDA_VER_MINOR=${CUDA_VERSION[1]}" | tee $0.out)
```
```suggestion
CUDA_VER_MAJOR=${CUDA_VERSION[0]}
CUDA_VER_MINOR=${CUDA_VERSION[1]}
echo -e "CUDA_VER_MAJOR=${CUDA_VERSION[0]}\nCUDA_VER_MINOR=${CUDA_VERSION[1]}" > "$0.out"
```
`eval` is dangerous.

tdulcet commented on .github/workflows/scripts/build_helper.sh:
```diff
@@ -0,0 +1,102 @@
+  COMPILER_VER="${CC_VSPROD}, $(echo $CC_VSINFO | grep InstallationVersion | cut -d':' -f2 | xargs)"
+else
+  COMPILER_VER="$(gcc --version | head -n1)"
+  OS_VER="$(grep PRETTY_NAME /etc/os-release | cut -d'=' -f2- | tr -d '"')"
```
```suggestion
  . /etc/os-release
  OS_VER=${PRETTY_NAME}
```
The `/etc/os-release` file is already a shell script, so just source it.

tdulcet commented on .github/workflows/scripts/build_helper.sh:
```diff
@@ -0,0 +1,102 @@
+# build_helper.sh.out which are later used by the actions workflow.
+# It also patches Makefiles to support building under Github Action
+# runners environment and compile kernels for every compute capability
+# device supported by the NVCC.
```
Please run this script through [ShellCheck](https://www.shellcheck.net/). Also, I would recommend adding a `set -e` somewhere near the top, so that it exits on error.

N-Storm replied:
I'm using VS Code with a ShellCheck plugin. Only 2 warnings which relates to adding quotes to a variables to avoid splits, but these variables shouldn't have such input anyways. So it's irrelevant for a script used only from GH Actions.

tdulcet commented on .github/workflows/scripts/build_helper.sh:
```diff
@@ -0,0 +1,102 @@
+fi
+
+echo 'Gathering version info on generic compiler and nvcc...'
+if [[ -x "$(which vswhere.exe)" ]]; then
```
`which` is not portable. Use `command -v` instead. Same below.

tdulcet replied:
It depends on if you want to limit this script to only CI/CD usage. In general, I usually avoid external commands in favor of the standard Bash builtins, whenever possible. See this for more information: https://www.shellcheck.net/wiki/SC2230

tdulcet commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,248 @@
+      - name: Prepare build archive with description
+        shell: bash
+        run: |
+          choco install -y --no-progress zip
+          zip -9 -j "${{ env.base_name }}.zip" *
```
The explicit zip should not be needed here, as one can upload multiple files into a single directory.

N-Storm replied:
Not sure what you mean here. These archives are created to be uploaded on release.

tdulcet replied:
I mainly meant that you would not need to install zip here, as you could instead wait until the separate `upload-release` job, which is run on Linux, where zip is already installed by default. That is [for example what I do](https://github.com/tdulcet/AutoPrimeNet/blob/ab1ad086ab33818b796698d61e2396228ec0c7e0/.github/workflows/cd.yml#L56-L73) when packaging the Windows standalone executables of AutoPrimeNet.

N-Storm replied:
Ah, I see now. But this step doesn't takes much time, and handling archiving on a upload-release job will require changes in the workflow which I'm not ready to implement right now. I'm leaving this as is, since it works fine and doesn't really impacts total build times. This could be done later.

tdulcet reviewed (commented):
Just some comments in case they are helpful. Feel free to ignore.

N-Storm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,248 @@
+      # MSVC 2022 installed on Windows 2022 Github Runner has PowerShell script for a Dev Shell.
+      - name: Build from sources (PowerShell with MSVC 2022)
+        if: ${{ matrix.sys.os == 'windows-2022' }}
+        shell: powershell
```
I think I can argue only with that one. It's 'pwsh.exe' that are default. Newer version, which sometimes produces other results. VS 2022 docs on DevShell suggests running that PS script with 'powershell.exe', so I'd rather stick to it. 
And quickly looking though others suggestions you made, I think I agree with all remaining. Thanks for your review. I'll take a closer look and apply or comment others once I'll have some time to check & test these.

N-Storm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,248 @@
+        sys:
+          # Specified version combination must exist as CUDA container image from NVidia: nvcr.io/nvidia/cuda:${{ matrix.sys.cuda_version }}-devel-${{ matrix.sys.ct_os }}
+          # Available versions can be found here: https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda/tags (note that only Ubuntus are supported by this action)
+          - { cuda_version: '12.8.0', ct_os: 'ubuntu24.04' }
+          - { cuda_version: '12.6.3', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.6.2', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.6.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.6.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.5.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.5.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.4.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.4.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.3.2', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.3.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.3.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.2.2', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.2.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.1.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.1.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.0.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '12.0.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '11.8.0', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '11.7.1', ct_os: 'ubuntu22.04' }
+          - { cuda_version: '11.6.2', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.6.1', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.5.2', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.4.3', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.3.1', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.2.2', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.1.1', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '11.0.3', ct_os: 'ubuntu20.04' }
+          - { cuda_version: '10.2', ct_os: 'ubuntu18.04' }
+          - { cuda_version: '9.2', ct_os: 'ubuntu18.04' }
+          - { cuda_version: '8.0', ct_os: 'ubuntu16.04' }
```
Thanks for pointing that out. Yes, I've decided to keep it filled on the PR and let maintainer decide which ones to remove (if any) later. But totally forgot about this decision while I was writing PR description.

tdulcet replied:
No problem. Keep in mind that all of the repositories in the Prime Search organization share the same pool of runners, so for example running a lot of jobs here could delay mfakto or Mlucas jobs from running (or vice versa).

N-Storm commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,248 @@
+          bash .github/workflows/scripts/build_helper.sh ${{ matrix.sys.cuda_version }}
+          cat .github/workflows/scripts/build_helper.sh.out >> $GITHUB_OUTPUT
+
+      # MSVC 2022 installed on Windows 2022 Github Runner has PowerShell script for a Dev Shell.
+      - name: Build from sources (PowerShell with MSVC 2022)
+        if: ${{ matrix.sys.os == 'windows-2022' }}
+        shell: powershell
+        run: |
+          & 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\Launch-VsDevShell.ps1' -Arch amd64 -HostArch amd64
+          cd "${{ github.workspace }}\src"
+          Copy-Item mfaktc.ini ..
+          make SHELL="powershell.exe" -f Makefile.win
+
+      # MSVC 2019 on Windows 2019 has similar script, but it doesn't allows setting arch & host_arch and defaults to x86 (32 bit) env.
+      # So we have to run bat file for the env, but Make uses PowerShell afterward because it's much better at handling long paths &
+      # quotes when invoked from Make.
+      - name: Build from sources (cmd.exe with MSVC 2019)
+        if: ${{ matrix.sys.os == 'windows-2019' }}
+        shell: cmd
+        run: |
+          "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvars64.bat" & cd src & copy mfaktc.ini .. & make SHELL="powershell.exe" -f Makefile.win
```
I've tried it. And a few others similar actions. But had issues working with GNU Make, quotes / spaces in paths issues. Maybe a workaround exists, but I didn't had any coming in reasonable time. We are running on the same Github runner images and I think even those hardcoded paths shouldn't be an issue. Making it more versatile where it's not required adds unnecessary complexity. And GNU Make with MSVC combo already tries to break things too easily.
I think VS 2022 can be run with a cmd init as well and then these steps can be combined into one. But I thought that it would be good to leave it like this because it's less prone to errors and as "boilerplate" for the next runner image version.

tdulcet replied:
Yeah, it looks like the `windows-2019` image is soon going to be replaced with the new `windows-2025` image. You might want to consider adding a job with the `windows-2025` image, to be prepared for the transition.

brubsby commented:
> Added patch to replace deprecated API calls as well. Let me know if want it out of this PR / v0.23.2, I'll exclude it and force-push the branch without it.

Sorry for the slow response, since we're making the same change in 0.24, I would lean towards not including it in 0.23.2, just on the off chance that it causes issues with anybody on Linux (as seems to happen sometimes), and doesn't seem to fix anything if I'm understanding correctly. Especially as we're aiming to stabilize 0.23 and move on to developing 0.24.0.

However it does seem low risk, and already in the PR, so I'll defer to your judgement. Worst case we can do a 0.23.3 that removes it if issues crop up, and with the CI/CD now, that's not as annoying as it once was.

brubsby reviewed (dismissed):
Feel free to make any changes suggested by Teal, and I'll re-review, but I'm okay for this to go in as-is. Might also be easier to cherry pick onto 0.24.0 if this PR is squashed, but I'm okay with non-squashed if you want to preserve the commit history.

N-Storm commented:
I was quite busy recently. Finally got time to return to this.
I've updated PR with a new commit. Removed deprecated CUDA calls patch and added most of the changes suggested by @tdulcet. I've squashed some of the latest commits into one, but keeping previous few commits with a history.

@brubsby, you can cherry-pick a range of commits in one run. See this link: https://stackoverflow.com/questions/9229301/git-cherry-pick-says-38c74d-is-a-merge-but-no-m-option-was-given

I'm waiting now for a build to complete on my fork to check for possible issues. Once it's all good and ready, I'll re-request review again and comment it here.

N-Storm commented:
Currently Windows builds are broken due to https://github.com/Jimver/cuda-toolkit/issues/382 / https://github.com/actions/runner-images/issues/11691
Hopefully it will be resolved soon. If not, workaround is possible by manually unpacking & installing CUDA instead, but getting installer fixed would be preferred solution.

N-Storm commented:
Confirmed to be a VS bug here: https://github.com/microsoft/STL/issues/5282

N-Storm commented:
It looks like a very easy workaround exists to install only some of the CUDA components. I've tested and looks like it's working, just need to add CUDA bin dir to PATH additionally (installing without broken "Visual Studio Integration" component which was used to set it). I'll test it a bit more and update PR if everything will be fine.

N-Storm commented:
I guess it's all good now, build passes: https://github.com/N-Storm/mfaktc/actions/runs/13642472654

And I've added CUDA 8.x, 9x and 10.x Windows builds as well. I hope now it should be ready to merge.

brubsby reviewed (approved):
Looks good! Thanks again for putting all this together.
